File: rendler.py
#!/usr/bin/env python

from collections import deque
import json
import os
import signal
import sys
import time
from threading import Thread

import mesos
import mesos_pb2
import results
import task_state
import export_dot

TASK_CPUS = 0.1
TASK_MEM = 32

# See the Mesos Framework Development Guide:
# http://mesos.apache.org/documentation/latest/app-framework-development-guide

class RenderingCrawler(mesos.Scheduler):
    def __init__(self, seedUrl, crawlExecutor, renderExecutor):
        print "RENDLER"
        print "======="
        print "seedUrl: [%s]\n" % seedUrl
        self.seedUrl = seedUrl
        self.crawlExecutor  = crawlExecutor
        self.renderExecutor = renderExecutor
        self.crawlQueue = deque([seedUrl])
        self.renderQueue = deque([seedUrl])
        self.processedURLs = set([seedUrl])
        self.crawlResults = set([])
        self.renderResults = {}
        self.tasksCreated  = 0
        self.tasksRunning = 0
        self.shuttingDown = False

    def registered(self, driver, frameworkId, masterInfo):
        print "Registered with framework ID [%s]" % frameworkId.value

    def makeTaskPrototype(self, offer):
        task = mesos_pb2.TaskInfo()
        tid = self.tasksCreated
        self.tasksCreated += 1
        task.task_id.value = str(tid)
        task.slave_id.value = offer.slave_id.value
        cpus = task.resources.add()
        cpus.name = "cpus"
        cpus.type = mesos_pb2.Value.SCALAR
        cpus.scalar.value = TASK_CPUS
        mem = task.resources.add()
        mem.name = "mem"
        mem.type = mesos_pb2.Value.SCALAR
        mem.scalar.value = TASK_MEM
        return task

    def makeCrawlTask(self, url, offer):
        task = self.makeTaskPrototype(offer)
        task.name = "crawl_%s" % task.task_id
        task.executor.MergeFrom(self.crawlExecutor)
        task.data = str(url)
        return task

    def makeRenderTask(self, url, offer):
        task = self.makeTaskPrototype(offer)
        task.name = "render_%s" % task.task_id
        task.executor.MergeFrom(self.renderExecutor)
        task.data = str(url)
        return task

    def printQueueStatistics(self):
        print "Crawl queue length: %d, Render queue length: %d, Running tasks: %d" % (
            len(self.crawlQueue), len(self.renderQueue), self.tasksRunning
        )

    def maxTasksForOffer(self, offer):
        count = 0
        cpus = next(rsc.scalar.value for rsc in offer.resources if rsc.name == "cpus")
        mem = next(rsc.scalar.value for rsc in offer.resources if rsc.name == "mem")
        while cpus >= TASK_CPUS and mem >= TASK_MEM:
            count += 1
            cpus -= TASK_CPUS
            mem -= TASK_MEM
        return count

    def resourceOffers(self, driver, offers):
        self.printQueueStatistics()

        for offer in offers:
            print "Got resource offer [%s]" % offer.id.value

            if self.shuttingDown:
                print "Shutting down: declining offer on [%s]" % offer.hostname
                driver.declineOffer(offer.id)
                continue

            maxTasks = self.maxTasksForOffer(offer)

            print "maxTasksForOffer: [%d]" % maxTasks

            tasks = []

            for i in range(maxTasks / 2):
                if self.crawlQueue:
                    crawlTaskUrl = self.crawlQueue.popleft()
                    task = self.makeCrawlTask(crawlTaskUrl, offer)
                    tasks.append(task)
                if self.renderQueue:
                    renderTaskUrl = self.renderQueue.popleft()
                    task = self.makeRenderTask(renderTaskUrl, offer)
                    tasks.append(task)

            if tasks:
                print "Accepting offer on [%s]" % offer.hostname
                driver.launchTasks(offer.id, tasks)
            else:
                print "Declining offer on [%s]" % offer.hostname
                driver.declineOffer(offer.id)

    def statusUpdate(self, driver, update):
        stateName = task_state.nameFor[update.state]
        print "Task [%s] is in state [%s]" % (update.task_id.value, stateName)

        if update.state == 1:  # Running
            self.tasksRunning += 1
        elif update.state == 3:  # Failed
            self.tasksRunning -= 1  # consider rescheduling the task
            print "Task [%s] failed with message \"%s\"" \
                % (update.task_id.value, update.message)
        elif self.tasksRunning > 0 and update.state > 1: # Terminal state
            self.tasksRunning -= 1

    def frameworkMessage(self, driver, executorId, slaveId, message):
        o = json.loads(message)

        if executorId.value == crawlExecutor.executor_id.value:
            result = results.CrawlResult(o['taskId'], o['url'], o['links'])
            for link in result.links:
                edge = (result.url, link)
                print "Appending [%s] to crawl results" % repr(edge)
                self.crawlResults.add(edge)
                if not link in self.processedURLs:
                    print "Enqueueing [%s]" % link
                    self.crawlQueue.append(link)
                    self.renderQueue.append(link)
                    self.processedURLs.add(link)

        elif executorId.value == renderExecutor.executor_id.value:
            result = results.RenderResult(o['taskId'], o['url'], o['imageUrl'])
            print "Appending [%s] to render results" % repr((result.url, result.imageUrl))
            self.renderResults[result.url] = result.imageUrl

def shutdown(signal, frame):
    print "Rendler is shutting down"
    rendler.shuttingDown = True
    while rendler.tasksRunning > 0:
        time.sleep(1)
    driver.stop()
    export_dot.dot(rendler.crawlResults, rendler.renderResults, "result.dot")
    print "Goodbye!"
    sys.exit(0)

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print "Usage: %s seedUrl mesosMasterUrl" % sys.argv[0]
        sys.exit(1)

    uris = [ "crawl_executor.py",
             "export_dot.py",
             "render.js",
             "render_executor.py",
             "results.py",
             "task_state.py" ]

    crawlExecutor = mesos_pb2.ExecutorInfo()
    crawlExecutor.executor_id.value = "crawl-executor"
    crawlExecutor.command.value = "python crawl_executor.py"

    for uri in uris:
        uri_proto = crawlExecutor.command.uris.add()
        uri_proto.value = "/home/vagrant/hostfiles/" + uri
        uri_proto.extract = False

    crawlExecutor.name = "Crawler"

    renderExecutor = mesos_pb2.ExecutorInfo()
    renderExecutor.executor_id.value = "render-executor"
    renderExecutor.command.value = "python render_executor.py --local"

    for uri in uris:
        uri_proto = renderExecutor.command.uris.add()
        uri_proto.value = "/home/vagrant/hostfiles/" + uri
        uri_proto.extract = False

    renderExecutor.name = "Renderer"

    framework = mesos_pb2.FrameworkInfo()
    framework.user = "" # Have Mesos fill in the current user.
    framework.name = "RENDLER"

    rendler = RenderingCrawler(sys.argv[1], crawlExecutor, renderExecutor)

    driver = mesos.MesosSchedulerDriver(rendler, framework, sys.argv[2])

    # driver.run() blocks; we run it in a separate thread
    def run_driver_async():
        status = 0 if driver.run() == mesos_pb2.DRIVER_STOPPED else 1
        driver.stop()
        sys.exit(status)

    Thread(target = run_driver_async, args = ()).start()

    print "(Listening for Ctrl-C)"
    signal.signal(signal.SIGINT, shutdown)
    while True: time.sleep(1)

File: go/src/rendler/rendler.go
package main

import (
	"code.google.com/p/goprotobuf/proto"
	"container/list"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"github.com/mesosphere/mesos-go/mesos"
	"os"
	"path/filepath"
)

const TASK_CPUS = 0.1
const TASK_MEM = 32.0

type Edge struct {
	From string
	To string
}

func (e Edge) String() string {
	return fmt.Sprintf("(%s, %s)", e.From, e.To)
}

func main() {

	crawlQueue := list.New() // list of string
	renderQueue := list.New() // list of string

	processedURLs := list.New() // list of string
	crawlResults := list.New() // list of CrawlEdge
	renderResults := make(map[string]string)

	seedUrl := flag.String("seed", "http://mesosphere.io", "The first URL to crawl")
	crawlQueue.PushBack(seedUrl)

	taskLimit := 5
	tasksCreated := 0

	exit := make(chan bool)

	master := flag.String("master", "localhost:5050", "Location of leading Mesos master")
	localMode := flag.Bool("--local", false, "If true, saves rendered web pages on local disk")
	flag.Parse()

	crawlCommand := "python crawl_executor.py"
	renderCommand := "python render_executor.py"

	if *localMode {
		renderCommand += " --local"
		crawlCommand += " --local"
	}

	rendlerArtifacts := executorURIs()

	log.Println("Executor URIs:")
	for i, artifact := range rendlerArtifacts {
		log.Printf("\t%d: %s", i, *artifact.Value)
	}

	crawlExecutor := &mesos.ExecutorInfo{
		ExecutorId: &mesos.ExecutorID{Value: proto.String("crawl-executor")},
		Command: &mesos.CommandInfo{
			Value: proto.String(crawlCommand),
			Uris: rendlerArtifacts,
		},
		Name:   proto.String("Crawler"),
		Source: proto.String("rendering-crawler"),
	}


	renderExecutor := &mesos.ExecutorInfo{
		ExecutorId: &mesos.ExecutorID{Value: proto.String("render-executor")},
		Command: &mesos.CommandInfo{
			Value: proto.String(renderCommand),
			Uris: rendlerArtifacts,
		},
		Name:   proto.String("Renderer"),
		Source: proto.String("rendering-crawler"),
	}

	makeTaskPrototype := func(offer mesos.Offer) *mesos.TaskInfo {
		taskId := tasksCreated
		tasksCreated++
		return &mesos.TaskInfo{
			TaskId: &mesos.TaskID{
				Value: proto.String(fmt.Sprintf("RENDLER-%d", taskId)),
			},
			SlaveId:  offer.SlaveId,
			Resources: []*mesos.Resource{
				mesos.ScalarResource("cpus", TASK_CPUS),
				mesos.ScalarResource("mem", TASK_MEM),
			},
		}
	}

	makeCrawlTask := func(url string, offer mesos.Offer) *mesos.TaskInfo {
		task := makeTaskPrototype(offer)
		task.Name = proto.String("CRAWL_" + *task.TaskId.Value)
		task.Executor = crawlExecutor
		task.Data = []byte(url)
		return task
	}

	makeRenderTask := func(url string, offer mesos.Offer) *mesos.TaskInfo {
		task := makeTaskPrototype(offer)
		task.Name = proto.String("RENDER_" + *task.TaskId.Value)
		task.Executor = renderExecutor
		task.Data = []byte(url)
		return task
	}


	maxTasksForOffer := func(offer mesos.Offer) int {
/*
        cpus = next(rsc.scalar.value for rsc in offer.resources if rsc.name == "cpus")
        mem = next(rsc.scalar.value for rsc in offer.resources if rsc.name == "mem")
*/
        count := 0
        cpus := 1.0 // TODO
        mem := 64.0 // TODO

        for cpus >= TASK_CPUS && mem >= TASK_MEM {
        	count++
        	cpus -= TASK_CPUS
        	mem -= TASK_MEM
        }

        return count
	}

	printQueueStatistics := func() {
		// TODO
	}

	driver := mesos.SchedulerDriver{
		Master: *master,
		Framework: mesos.FrameworkInfo{
			Name: proto.String("RENDLER"),
			User: proto.String(""),
		},

		Scheduler: &mesos.Scheduler{

			Registered: func(
				driver *mesos.SchedulerDriver,
				frameworkId mesos.FrameworkID,
				masterInfo mesos.MasterInfo) {
				// TODO
			},

			ResourceOffers: func(driver *mesos.SchedulerDriver, offers []mesos.Offer) {
				printQueueStatistics()

				for _, offer := range offers {
					maxTasks := maxTasksForOffer(offer)
					log.Printf("maxTasksForOffer: [%d]", maxTasks)

					tasks := []mesos.TaskInfo{}

					for i := 0; i < maxTasksForOffer(offer) / 2; i++ {
						if crawlQueue.Front() != nil {
							url := crawlQueue.Front().Value.(*string)
							crawlQueue.Remove(crawlQueue.Front())
							task := makeCrawlTask(*url, offer)
							tasks = append(tasks, *task)
						}
						if renderQueue.Front() != nil {
							url := renderQueue.Front().Value.(*string)
							renderQueue.Remove(renderQueue.Front())
							task := makeRenderTask(*url, offer)
							tasks = append(tasks, *task)
						}
					}

					if len(tasks) == 0 {
						driver.DeclineOffer(offer.Id)
					} else {
						driver.LaunchTasks(offer.Id, tasks)
					}
				}
			},

			StatusUpdate: func(driver *mesos.SchedulerDriver, status mesos.TaskStatus) {
				log.Printf("Received task status: " + *status.Message)

				if *status.State == mesos.TaskState_TASK_FINISHED {
					taskLimit--
					if taskLimit <= 0 {
						exit <- true
					}
				}
			},

			FrameworkMessage: func(
				driver *mesos.SchedulerDriver,
				executorId mesos.ExecutorID,
				slaveId mesos.SlaveID,
				message string) {

				switch executorId.Value {
				case crawlExecutor.ExecutorId.Value:
					log.Print("Received framework message from crawler")
					var result CrawlResult
					err := json.Unmarshal([]byte(message), &result)
					if err != nil {
						log.Printf("Error deserializing CrawlResult: [%s]", err)
					} else {
						for _, link := range result.Links {
							edge := Edge{From: result.URL, To: link,}
							log.Printf("Appending [%s] to crawl results", edge)
							crawlResults.PushBack(edge)

							alreadyProcessed := false
							for e := processedURLs.Front(); e != nil && !alreadyProcessed; e = e.Next() {
								processedURL := e.Value.(string)
								if link == processedURL {
									alreadyProcessed = true
								}
							}

							if !alreadyProcessed {
								log.Printf("Enqueueing [%s]", link)
								crawlQueue.PushBack(link)
								renderQueue.PushBack(link)
								processedURLs.PushBack(link)
							}
						}
					}

				case renderExecutor.ExecutorId.Value:
					log.Printf("Received framework message from renderer")
					var result RenderResult
					err := json.Unmarshal([]byte(message), &result)
					if err != nil {
						log.Printf("Error deserializing RenderResult: [%s]", err)
					} else {
						log.Printf(
							"Appending [%s] to render results",
							Edge{From: result.URL, To: result.ImageURL})
						renderResults[result.URL] = result.ImageURL
					}

				default:
					log.Printf("Received a framework message from some unknown source")
				}
			},
		},
	}

	driver.Init()
	defer driver.Destroy()

	driver.Start()
	<-exit
	driver.Stop(false)
}

func executorURIs() []*mesos.CommandInfo_URI {
	basePath, err := filepath.Abs(filepath.Dir(os.Args[0]) + "/../../..")
	if err != nil {
		log.Fatal("Failed to find the path to RENDLER")
	}
	baseURI := fmt.Sprintf("file://%s/", basePath)

	pathToURI := func(path string, extract bool) *mesos.CommandInfo_URI{
		return &mesos.CommandInfo_URI{
			Value: &path,
			Extract: &extract,
		}
	}

	return []*mesos.CommandInfo_URI{
		pathToURI(baseURI + "crawl_executor.py", false),
		pathToURI(baseURI + "render.js", false),
		pathToURI(baseURI + "render_executor.py", false),
		pathToURI(baseURI + "results.py", false),
		pathToURI(baseURI + "task_state.py", false),
	}
}
